cmd/task/list: unhandled panic

```
./funnel task list
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x72e404]

goroutine 1 [running]:
net/http.(*Request).WithContext(0x0, 0x1c97b60, 0xc420010250, 0x1c97b60)
	/usr/local/go/src/net/http/request.go:332 +0x44
github.com/ohsu-comp-bio/funnel/client.(*Client).ListTasks(0xc4204782a0, 0x7f47d9c1a050, 0xc420010250, 0xc4201e5c50, 0xc420010250, 0xc4203d8d01, 0x6)
	/Users/buchanae/src/github.com/ohsu-comp-bio/funnel/client/client.go:78 +0x221
github.com/ohsu-comp-bio/funnel/cmd/task.List(0xc42000a14e, 0x10, 0x13b9a22, 0x5, 0x0, 0x0, 0x0, 0x1c8d6a0, 0xc42000c018, 0xc4201e5d18, ...)
	/Users/buchanae/src/github.com/ohsu-comp-bio/funnel/cmd/task/list.go:30 +0x1b8
github.com/ohsu-comp-bio/funnel/cmd/task.newCommandHooks.func3(0xc420476d80, 0x1d50a30, 0x0, 0x0, 0x0, 0x0)
	/Users/buchanae/src/github.com/ohsu-comp-bio/funnel/cmd/task/task.go:64 +0xfc
github.com/ohsu-comp-bio/funnel/vendor/github.com/spf13/cobra.(*Command).execute(0xc420476d80, 0x1d50a30, 0x0, 0x0, 0xc420476d80, 0x1d50a30)
	/Users/buchanae/src/github.com/ohsu-comp-bio/funnel/vendor/github.com/spf13/cobra/command.go:698 +0x47b
github.com/ohsu-comp-bio/funnel/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x1d286a0, 0x1d29e00, 0x0, 0x0)
	/Users/buchanae/src/github.com/ohsu-comp-bio/funnel/vendor/github.com/spf13/cobra/command.go:783 +0x349
github.com/ohsu-comp-bio/funnel/vendor/github.com/spf13/cobra.(*Command).Execute(0x1d286a0, 0xc420000340, 0xc420000340)
	/Users/buchanae/src/github.com/ohsu-comp-bio/funnel/vendor/github.com/spf13/cobra/command.go:736 +0x2b
main.main()
	/Users/buchanae/src/github.com/ohsu-comp-bio/funnel/main.go:10 +0x2d
```

--- a/client/client.go
+++ b/client/client.go
@@ -26,12 +26,11 @@ func NewClient(address string) (*Client, error) {
 	endpoint := re.ReplaceAllString(address, "$1$2")
 
 	reScheme := regexp.MustCompile("^.+://")
-	switch reScheme.MatchString(endpoint) {
-	case true:
+	if reScheme.MatchString(endpoint) {
 		if !strings.HasPrefix(endpoint, "http") {
 			return nil, fmt.Errorf("invalid protocol: '%s'; expected: 'http://' or 'https://'", reScheme.FindString(endpoint))
 		}
-	case false:
+	} else {
 		endpoint = "http://" + endpoint
 	}
 

--- a/client/client.go
+++ b/client/client.go
@@ -12,27 +12,37 @@ import (
 	"net/url"
 	"os"
 	"regexp"
+	"strings"
 	"time"
 )
 
 // NewClient returns a new HTTP client for accessing
 // Create/List/Get/Cancel Task endpoints. "address" is the address
 // of the TES server.
-func NewClient(address string) *Client {
+func NewClient(address string) (*Client, error) {
 	password := os.Getenv("FUNNEL_SERVER_PASSWORD")
 
-	// Strip trailing slash. A quick and dirty fix.
-	re := regexp.MustCompile("/+$")
-	address = re.ReplaceAllString(address, "")
+	re := regexp.MustCompile("^(.+://)?(.[^/]+)(.+)?$")
+	endpoint := re.ReplaceAllString(address, "$1$2")
+
+	reScheme := regexp.MustCompile("^.+://")
+	switch reScheme.MatchString(endpoint) {
+	case true:
+		if !strings.HasPrefix(endpoint, "http") {
+			return nil, fmt.Errorf("invalid protocol: '%s'; expected: 'http://' or 'https://'", reScheme.FindString(endpoint))
+		}
+	case false:
+		endpoint = "http://" + endpoint
+	}
 
 	return &Client{
-		address: address,
+		address: endpoint,
 		client: &http.Client{
-			Timeout: 30 * time.Second,
+			Timeout: 60 * time.Second,
 		},
 		Marshaler: &tes.Marshaler,
 		Password:  password,
-	}
+	}, nil
 }
 
 // Client represents the HTTP Task client.

--- a/client/client_test.go
+++ b/client/client_test.go
@@ -23,7 +23,10 @@ func testServer(mux http.Handler) *httptest.Server {
 }
 
 func TestAddressTrailingSlash(t *testing.T) {
-	c := NewClient("http://funnel.com:8000/")
+	c, err := NewClient("http://funnel.com:8000/")
+	if err != nil {
+		t.Fatal(err)
+	}
 	if c.address != "http://funnel.com:8000" {
 		t.Error("Expected trailing slash to be stripped")
 	}
@@ -43,7 +46,10 @@ func TestGetTask(t *testing.T) {
 	defer ts.Close()
 
 	// Make test client call
-	c := NewClient("http://localhost:20001")
+	c, err := NewClient("http://localhost:20001")
+	if err != nil {
+		t.Fatal(err)
+	}
 	body, err := c.GetTask(context.Background(), &tes.GetTaskRequest{
 		Id:   "test-id",
 		View: tes.TaskView_MINIMAL,
@@ -71,7 +77,10 @@ func TestGetTaskTrailingSlash(t *testing.T) {
 	defer ts.Close()
 
 	// Make test client call
-	c := NewClient("http://localhost:20001")
+	c, err := NewClient("http://localhost:20001")
+	if err != nil {
+		t.Fatal(err)
+	}
 	body, err := c.GetTask(context.Background(), &tes.GetTaskRequest{
 		Id:   "test-id",
 		View: tes.TaskView_MINIMAL,
@@ -96,10 +105,13 @@ func TestClientTimeout(t *testing.T) {
 	ts := testServer(mux)
 	defer ts.Close()
 
-	c := NewClient("http://localhost:20001")
+	c, err := NewClient("http://localhost:20001")
+	if err != nil {
+		t.Fatal(err)
+	}
 	c.client.Timeout = 1 * time.Second
 
-	_, err := c.GetTask(context.Background(), &tes.GetTaskRequest{
+	_, err = c.GetTask(context.Background(), &tes.GetTaskRequest{
 		Id:   "test-id",
 		View: tes.TaskView_MINIMAL,
 	})

--- a/cmd/node/node.go
+++ b/cmd/node/node.go
@@ -61,6 +61,7 @@ func newCommandHooks() (*cobra.Command, *hooks) {
 	run := &cobra.Command{
 		Use:   "run",
 		Short: "Run a Funnel node.",
+		Args:  cobra.NoArgs,
 		RunE: func(cmd *cobra.Command, args []string) error {
 			return hooks.Run(conf)
 		},

--- a/cmd/run/cmd.go
+++ b/cmd/run/cmd.go
@@ -73,10 +73,15 @@ func Run(args []string) error {
 		return err
 	}
 
+	cli, err := client.NewClient(vals.server)
+	if err != nil {
+		return err
+	}
+
 	// TES HTTP client
 	tg := taskGroup{
 		printTask: vals.printTask,
-		client:    client.NewClient(vals.server),
+		client:    cli,
 	}
 
 	// Scatter all vals into tasks

--- a/cmd/server/server.go
+++ b/cmd/server/server.go
@@ -57,6 +57,7 @@ func newCommandHooks() (*cobra.Command, *hooks) {
 	run := &cobra.Command{
 		Use:   "run",
 		Short: "Runs a Funnel server.",
+		Args:  cobra.NoArgs,
 		RunE: func(cmd *cobra.Command, args []string) error {
 			return hooks.Run(context.Background(), conf)
 		},

--- a/cmd/task/cancel.go
+++ b/cmd/task/cancel.go
@@ -11,7 +11,11 @@ import (
 // Cancel runs the "task cancel" CLI command, which connects to the server,
 // calls CancelTask() on each ID, and writes output to the given writer.
 func Cancel(server string, ids []string, writer io.Writer) error {
-	cli := client.NewClient(server)
+	cli, err := client.NewClient(server)
+	if err != nil {
+		return err
+	}
+
 	res := []string{}
 
 	for _, taskID := range ids {

--- a/cmd/task/create.go
+++ b/cmd/task/create.go
@@ -14,7 +14,11 @@ import (
 // calling CreateTask, and writing output to the given writer.
 // Tasks are loaded from the "files" arg. "files" are file paths to JSON objects.
 func Create(server string, files []string, writer io.Writer) error {
-	cli := client.NewClient(server)
+	cli, err := client.NewClient(server)
+	if err != nil {
+		return err
+	}
+
 	res := []string{}
 
 	for _, taskFile := range files {

--- a/cmd/task/get.go
+++ b/cmd/task/get.go
@@ -12,7 +12,11 @@ import (
 // calls GetTask for each ID, requesting the given task view, and writes
 // output to the given writer.
 func Get(server string, ids []string, taskView string, w io.Writer) error {
-	cli := client.NewClient(server)
+	cli, err := client.NewClient(server)
+	if err != nil {
+		return err
+	}
+
 	res := []string{}
 
 	view, ok := tes.TaskView_value[taskView]

--- a/cmd/task/get.go
+++ b/cmd/task/get.go
@@ -19,9 +19,9 @@ func Get(server string, ids []string, taskView string, w io.Writer) error {
 
 	res := []string{}
 
-	view, ok := tes.TaskView_value[taskView]
-	if !ok {
-		return fmt.Errorf("Unknown task view: %s", taskView)
+	view, err := getTaskView(taskView)
+	if err != nil {
+		return err
 	}
 
 	for _, taskID := range ids {

--- a/cmd/task/list.go
+++ b/cmd/task/list.go
@@ -18,9 +18,9 @@ func List(server, taskView, pageToken string, pageSize uint32, all bool, writer 
 		return err
 	}
 
-	view, ok := tes.TaskView_value[taskView]
-	if !ok {
-		return fmt.Errorf("Unknown task view: %s", taskView)
+	view, err := getTaskView(taskView)
+	if err != nil {
+		return err
 	}
 
 	output := &tes.ListTasksResponse{}

--- a/cmd/task/list.go
+++ b/cmd/task/list.go
@@ -13,7 +13,10 @@ import (
 // client-side using the "taskState" and "taskName" args. Output is written
 // to the given writer.
 func List(server, taskView, pageToken string, pageSize uint32, all bool, writer io.Writer) error {
-	cli := client.NewClient(server)
+	cli, err := client.NewClient(server)
+	if err != nil {
+		return err
+	}
 
 	view, ok := tes.TaskView_value[taskView]
 	if !ok {

--- a/cmd/task/task.go
+++ b/cmd/task/task.go
@@ -2,9 +2,11 @@ package task
 
 import (
 	"fmt"
+	"github.com/ohsu-comp-bio/funnel/proto/tes"
 	"github.com/spf13/cobra"
 	"io"
 	"os"
+	"strings"
 )
 
 // NewCommand returns the "task" subcommands.
@@ -54,37 +56,35 @@ func newCommandHooks() (*cobra.Command, *hooks) {
 		pageToken string
 		pageSize  uint32
 		listAll   bool
+		listView  string
 	)
-	listView := choiceVar{val: "BASIC"}
 
 	list := &cobra.Command{
 		Use:   "list",
 		Short: "List all tasks.",
 		RunE: func(cmd *cobra.Command, args []string) error {
-			return h.List(tesServer, listView.val, pageToken, pageSize, listAll, cmd.OutOrStdout())
+			return h.List(tesServer, listView, pageToken, pageSize, listAll, cmd.OutOrStdout())
 		},
 	}
 
 	lf := list.Flags()
-	listView.AddChoices("BASIC", "MINIMAL", "FULL")
-	lf.VarP(&listView, "view", "v", "Task view")
+	lf.StringVarP(&listView, "view", "v", "basic", "Task view")
 	lf.StringVarP(&pageToken, "page-token", "p", pageToken, "Page token")
 	lf.Uint32VarP(&pageSize, "page-size", "s", pageSize, "Page size")
 	lf.BoolVar(&listAll, "all", listAll, "List all tasks")
 
-	getView := choiceVar{val: "FULL"}
+	var getView string
 	get := &cobra.Command{
 		Use:   "get [taskID ...]",
 		Short: "Get one or more tasks by ID.",
 		Args:  cobra.MinimumNArgs(1),
 		RunE: func(cmd *cobra.Command, args []string) error {
-			return h.Get(tesServer, args, getView.val, cmd.OutOrStdout())
+			return h.Get(tesServer, args, getView, cmd.OutOrStdout())
 		},
 	}
 
 	gf := get.Flags()
-	getView.AddChoices("BASIC", "MINIMAL", "FULL")
-	gf.VarP(&getView, "view", "v", "Task view")
+	gf.StringVarP(&getView, "view", "v", "full", "Task view")
 
 	cancel := &cobra.Command{
 		Use:   "cancel [taskID ...]",
@@ -116,36 +116,13 @@ type hooks struct {
 	Wait   func(server string, ids []string) error
 }
 
-type choiceVar struct {
-	choices map[string]bool
-	val     string
-}
-
-func (c *choiceVar) AddChoices(choices ...string) {
-	if c.choices == nil {
-		c.choices = map[string]bool{}
+func getTaskView(taskView string) (int32, error) {
+	taskView = strings.ToUpper(taskView)
+	var view int32
+	var ok bool
+	view, ok = tes.TaskView_value[taskView]
+	if !ok {
+		return view, fmt.Errorf("Unknown task view: %s. Valid task views: ['basic', 'minimal', 'full']", taskView)
 	}
-	for _, choice := range choices {
-		c.choices[choice] = true
-	}
-}
-
-func (c *choiceVar) String() string {
-	return c.val
-}
-
-func (c *choiceVar) Set(v string) error {
-	if _, ok := c.choices[v]; !ok {
-		return fmt.Errorf("invalid choice: %s", v)
-	}
-	c.val = v
-	return nil
-}
-
-func (c *choiceVar) Get() interface{} {
-	return c.val
-}
-
-func (c *choiceVar) Type() string {
-	return "string"
+	return view, nil
 }

--- a/cmd/task/task_test.go
+++ b/cmd/task/task_test.go
@@ -28,8 +28,8 @@ func TestGetDefaultView(t *testing.T) {
 	cmd, h := newCommandHooks()
 
 	h.Get = func(server string, ids []string, view string, w io.Writer) error {
-		if view != "FULL" {
-			t.Errorf("expected default FULL view, got '%s'", view)
+		if view != "full" {
+			t.Errorf("expected default 'full' view, got '%s'", view)
 		}
 		return nil
 	}

--- a/cmd/task/wait.go
+++ b/cmd/task/wait.go
@@ -9,5 +9,10 @@ import (
 // calling GetTask() for each ID, and blocking until the tasks have
 // reached a terminal state.
 func Wait(server string, ids []string) error {
-	return client.NewClient(server).WaitForTask(context.Background(), ids...)
+	cli, err := client.NewClient(server)
+	if err != nil {
+		return err
+	}
+
+	return cli.WaitForTask(context.Background(), ids...)
 }

--- a/cmd/termdash/cursor.go
+++ b/cmd/termdash/cursor.go
@@ -16,10 +16,15 @@ type GridCursor struct {
 	isScrolling bool // toggled when actively scrolling
 }
 
-func NewGridCursor(tesHTTPServerAddress string) *GridCursor {
-	return &GridCursor{
-		tSource: NewTaskSource(tesHTTPServerAddress, 100),
+func NewGridCursor(tesHTTPServerAddress string) (*GridCursor, error) {
+	ts, err := NewTaskSource(tesHTTPServerAddress, 100)
+	if err != nil {
+		return nil, err
 	}
+
+	return &GridCursor{
+		tSource: ts,
+	}, nil
 }
 
 func (gc *GridCursor) Len() int { return len(gc.filtered) }

--- a/cmd/termdash/tasksource.go
+++ b/cmd/termdash/tasksource.go
@@ -25,11 +25,11 @@ type TaskSource struct {
 	lock     sync.RWMutex
 }
 
-func NewTaskSource(tesHTTPServerAddress string, pageSize uint32) *TaskSource {
+func NewTaskSource(tesHTTPServerAddress string, pageSize uint32) (*TaskSource, error) {
 	// init funnel http client
 	cli, err := client.NewClient(tesHTTPServerAddress)
 	if err != nil {
-		panic(err)
+		return nil, err
 	}
 	ts := &TaskSource{
 		client:   cli,
@@ -37,7 +37,7 @@ func NewTaskSource(tesHTTPServerAddress string, pageSize uint32) *TaskSource {
 		lock:     sync.RWMutex{},
 	}
 	ts.tasks, _ = ts.listTasks(false, false)
-	return ts
+	return ts, nil
 }
 
 func (ts *TaskSource) listTasks(previous, next bool) (TaskWidgets, error) {

--- a/cmd/termdash/tasksource.go
+++ b/cmd/termdash/tasksource.go
@@ -27,7 +27,10 @@ type TaskSource struct {
 
 func NewTaskSource(tesHTTPServerAddress string, pageSize uint32) *TaskSource {
 	// init funnel http client
-	cli := client.NewClient(tesHTTPServerAddress)
+	cli, err := client.NewClient(tesHTTPServerAddress)
+	if err != nil {
+		panic(err)
+	}
 	ts := &TaskSource{
 		client:   cli,
 		pageSize: pageSize,

--- a/cmd/termdash/termdash.go
+++ b/cmd/termdash/termdash.go
@@ -7,35 +7,39 @@ import (
 	"github.com/ohsu-comp-bio/funnel/cmd/termdash/config"
 	"github.com/ohsu-comp-bio/funnel/cmd/termdash/widgets"
 	"github.com/spf13/cobra"
+	"os"
 )
 
 var (
-	tesServer string
-	cursor    *GridCursor
-	cGrid     *compact.Grid
-	header    *widgets.TermDashHeader
+	defaultTesServer string = "http://localhost:8000"
+	tesServer        string = defaultTesServer
+	cursor           *GridCursor
+	cGrid            *compact.Grid
+	header           *widgets.TermDashHeader
 )
 
 // Cmd represents the worker command
 var Cmd = &cobra.Command{
 	Use:   "dashboard",
 	Short: "Start a Funnel dashboard in your terminal.",
-	Run: func(cmd *cobra.Command, args []string) {
-		termdash(tesServer)
+	PersistentPreRun: func(cmd *cobra.Command, args []string) {
+		if tesServer == defaultTesServer {
+			if val := os.Getenv("FUNNEL_SERVER"); val != "" {
+				tesServer = val
+			}
+		}
+	},
+	Args: cobra.NoArgs,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return termdash(tesServer)
 	},
 }
 
 func init() {
-	Cmd.Flags().StringVarP(&tesServer, "server", "S", "http://localhost:8000", "")
+	Cmd.Flags().StringVarP(&tesServer, "server", "S", tesServer, "")
 }
 
-func termdash(tesHTTPServerAddress string) {
-	defer func() {
-		if r := recover(); r != nil {
-			fmt.Println("An error occurred:\n", r)
-		}
-	}()
-
+func termdash(tesHTTPServerAddress string) error {
 	// init global config
 	config.Init()
 
@@ -43,19 +47,23 @@ func termdash(tesHTTPServerAddress string) {
 	ui.ColorMap = colorMap
 
 	if err := ui.Init(); err != nil {
-		panic(fmt.Errorf("Error initializing termdash UI: %v", err))
+		return fmt.Errorf("Error initializing termdash UI: %v", err)
 	}
 	defer Shutdown()
 
 	// init grid, cursor, header
 	header = widgets.NewTermDashHeader()
-	cursor = NewGridCursor(tesHTTPServerAddress)
 	cGrid = compact.NewGrid()
+	var err error
+	cursor, err = NewGridCursor(tesHTTPServerAddress)
+	if err != nil {
+		return fmt.Errorf("error initializing the grid cursor: %v", err)
+	}
 
 	for {
 		exit := Display()
 		if exit {
-			return
+			return nil
 		}
 	}
 }

--- a/cmd/worker/worker.go
+++ b/cmd/worker/worker.go
@@ -75,6 +75,7 @@ func newCommandHooks() (*cobra.Command, *hooks) {
 	run := &cobra.Command{
 		Use:   "run",
 		Short: "Run a task directly, bypassing the server.",
+		Args:  cobra.NoArgs,
 		RunE: func(cmd *cobra.Command, args []string) error {
 			if taskID == "" {
 				return fmt.Errorf("no taskID was provided")

--- a/tests/funnel_utils.go
+++ b/tests/funnel_utils.go
@@ -68,7 +68,7 @@ func NewFunnel(conf config.Config) *Funnel {
 
 	cli, err := client.NewClient(conf.Server.HTTPAddress())
 	if err != nil {
-		return err
+		panic(err)
 	}
 
 	dcli, derr := dockerutil.NewDockerClient()

--- a/tests/funnel_utils.go
+++ b/tests/funnel_utils.go
@@ -66,6 +66,11 @@ type Funnel struct {
 func NewFunnel(conf config.Config) *Funnel {
 	conf = config.EnsureServerProperties(conf)
 
+	cli, err := client.NewClient(conf.Server.HTTPAddress())
+	if err != nil {
+		return err
+	}
+
 	dcli, derr := dockerutil.NewDockerClient()
 	if derr != nil {
 		panic(derr)
@@ -77,7 +82,7 @@ func NewFunnel(conf config.Config) *Funnel {
 	}
 
 	return &Funnel{
-		HTTP:       client.NewClient(conf.Server.HTTPAddress()),
+		HTTP:       cli,
 		Docker:     dcli,
 		Conf:       conf,
 		StorageDir: conf.Worker.Storage.Local.AllowedDirs[0],

